common/persistence/serialization/task_serializer.go

```diff
@@ -0,0 +1,890 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package serialization
+
+import (
+	"fmt"
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/serviceerror"
+
+	enumsspb "go.temporal.io/server/api/enums/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"go.temporal.io/server/common/tasks"
+)
+
+type (
+	TaskSerializer struct {
+		namespaceID string
+		workflowID  string
+		runID       string
+	}
+)
+
+func NewTaskSerializer(
+	namespaceID string,
+	workflowID string,
+	runID string,
+) *TaskSerializer {
+	return &TaskSerializer{
+		namespaceID: namespaceID,
+		workflowID:  workflowID,
+		runID:       runID,
+	}
+}
+
+func (s *TaskSerializer) SerializeTransferTasks(
+	taskSlice []tasks.Task,
+) ([]commonpb.DataBlob, error) {
+	blobSlice := make([]commonpb.DataBlob, len(taskSlice))
+	for index, task := range taskSlice {
+		var transferTask *persistencespb.TransferTaskInfo
+		switch task.GetType() {
```

Comment:
Consider switch on `task` type and check that `GetType()` value match it:
```suggestion
		switch task:=task.(type) {
		case *tasks.WorkflowTask:
		   if task.GetType() != enumsspb.TASK_TYPE_TRANSFER_WORKFLOW_TASK {
		       return nil, serviceerror.NewInternal(fmt.Sprintf("Unexpected transfer task type: %v", task.GetType()))
		   }
		   transferTask = s.transferWorkflowTaskToProto(task)
```